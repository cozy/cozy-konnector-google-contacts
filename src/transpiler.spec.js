const transpile = require('./transpiler')

describe('Transpile from google-contacts to io.cozy-contacts', () => {
  it('Simple test', () => {
    const source = require('../sample/alanturing.json')

    const transpiled = transpile.toCozy(source)
    expect(transpiled).toMatchSnapshot()
  })

  it('complete tests with `sample/*.json', () => {
    const source = require('../sample/google-contacts.json')

    const transpiled = source.map(row => ({
      doc: transpile.toCozy(row)
    }))
    expect(transpiled).toMatchSnapshot()
  })
})

const johnDoeContact = {
  id: '9ecfbf4b-20e7-4bac-87f1-eea53350857d',
  _id: '9ecfbf4b-20e7-4bac-87f1-eea53350857d',
  _type: 'io.cozy.contacts',
  _rev: '1-19c313536e8b27473aa26bf105b03269',
  address: [
    {
      formattedAddress: '94 Hinton Road 05034 Fresno, Singapore',
      label: 'Home',
      type: 'home',
      primary: true
    },
    {
      street: '426 Runolfsson Knolls',
      city: 'Port Easter',
      country: 'Cocos (Keeling) Islands',
      postcode: '84573',
      label: 'Work',
      type: 'work'
    }
  ],
  email: [
    {
      address: 'john.doe@posteo.net',
      type: 'personal',
      primary: false
    },
    {
      address: 'john.doe@cozycloud.cc',
      primary: true
    }
  ],
  birthday: '1999-5-1',
  company: 'Cozy cloud',
  cozy: [
    {
      label: 'MyCozy',
      primary: true,
      url: 'https://johndoe.mycozy.cloud'
    }
  ],
  fullname: 'John Doe',
  name: {
    givenName: 'John',
    familyName: 'Doe'
  },
  metadata: {
    cozy: true,
    version: 1
  },
  note:
    'Atque cupiditate saepe omnis quos ut molestiae labore voluptates omnis.',
  phone: [
    {
      label: 'Work',
      number: '+33 (2)0 90 00 54 04',
      type: 'work',
      primary: true
    },
    {
      number: '+33 6 77 11 22 33',
      primary: false
    }
  ]
}

describe('Transpile from io.cozy.contacts to google-contacts', () => {
  it('should transpile a cozy contact to a google person', () => {
    const expected = {
      names: [
        {
          givenName: 'John',
          familyName: 'Doe'
        }
      ],
      emailAddresses: [
        {
          type: 'personal',
          value: 'john.doe@posteo.net',
          metadata: { primary: false }
        },
        {
          type: undefined,
          value: 'john.doe@cozycloud.cc',
          metadata: { primary: true }
        }
      ],
      phoneNumbers: [
        {
          value: '+33 (2)0 90 00 54 04',
          formattedType: 'Work',
          type: 'work',
          metadata: {
            primary: true
          }
        },
        {
          value: '+33 6 77 11 22 33',
          formattedType: undefined,
          type: undefined,
          metadata: {
            primary: false
          }
        }
      ],
      addresses: [
        {
          streetAddress: undefined,
          city: undefined,
          country: undefined,
          postalCode: undefined,
          formattedValue: '94 Hinton Road 05034 Fresno, Singapore',
          formattedType: 'Home',
          type: 'home',
          metadata: { primary: true }
        },
        {
          streetAddress: '426 Runolfsson Knolls',
          city: 'Port Easter',
          country: 'Cocos (Keeling) Islands',
          postalCode: '84573',
          formattedType: 'Work',
          formattedValue: undefined,
          type: 'work',
          metadata: { primary: false }
        }
      ],
      birthdays: [
        {
          metadata: {
            primary: true
          },
          date: { year: '1999', month: '5', day: '1' }
        }
      ],
      organizations: [
        {
          name: 'Cozy cloud'
        }
      ],
      biographies: [
        {
          contentType: 'TEXT_PLAIN',
          value:
            'Atque cupiditate saepe omnis quos ut molestiae labore voluptates omnis.'
        }
      ]
    }

    const actual = transpile.toGoogle(johnDoeContact)
    expect(actual).toEqual(expected)
  })
})
