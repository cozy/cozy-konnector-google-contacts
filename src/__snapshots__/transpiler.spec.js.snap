// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transpile from google-contacts to io.cozy-contacts should transpile alan turing 1`] = `
Object {
  "address": Array [
    Object {
      "city": "Hampton",
      "country": "UK",
      "label": "Work",
      "postcode": undefined,
      "primary": true,
      "street": "78 High Street",
      "type": "work",
    },
  ],
  "birthday": "1912-06-23",
  "company": "National Physical Laboratory",
  "email": Array [
    Object {
      "address": "alan.turing@npl.com",
      "label": "Work",
      "primary": true,
      "type": "work",
    },
    Object {
      "address": "alan.turing@wanadoo.fr",
      "label": "Home",
      "primary": false,
      "type": "home",
    },
  ],
  "metadata": Object {
    "google": Object {
      "metadata": Object {
        "objectType": "PERSON",
        "sources": Array [
          Object {
            "etag": "#H6QA7yWPAHE=",
            "id": "7486d2d60ee594e7",
            "type": "CONTACT",
            "updateTime": "2018-04-03T15:16:02.276Z",
          },
        ],
      },
    },
    "version": 1,
  },
  "name": Object {
    "familyName": "Turing",
    "givenName": "Alan",
  },
  "note": "Doloribus animi corporis et qui error cumque culpa autem velit.",
  "phone": Array [
    Object {
      "label": "Home",
      "number": "+44000555000",
      "primary": true,
      "type": "home",
    },
    Object {
      "label": "Work",
      "number": "+44 55 5000 555",
      "primary": false,
      "type": "work",
    },
  ],
  "vendorId": "people/c8396630372093629671",
}
`;

exports[`Transpile from google-contacts to io.cozy-contacts should transpile all contacts from sample/*.json 1`] = `
Array [
  Object {
    "doc": Object {
      "address": Array [
        Object {
          "city": undefined,
          "country": undefined,
          "label": undefined,
          "postcode": undefined,
          "primary": true,
          "street": "94 Crawford Street, 35599 Fairfield Kenya",
          "type": undefined,
        },
      ],
      "birthday": undefined,
      "company": undefined,
      "email": Array [
        Object {
          "address": "russell.bourquin@protheon.org",
          "label": undefined,
          "primary": true,
          "type": undefined,
        },
      ],
      "metadata": Object {
        "google": Object {
          "metadata": Object {
            "objectType": "PERSON",
            "sources": Array [
              Object {
                "etag": "#dIt0zEGXNVs=",
                "id": "23ddfe4b8c9b2ec1",
                "type": "CONTACT",
                "updateTime": "2018-03-20T11:03:08.585001Z",
              },
            ],
          },
        },
        "version": 1,
      },
      "name": Object {
        "familyName": undefined,
        "givenName": "Protheon",
      },
      "note": undefined,
      "phone": Array [
        Object {
          "label": undefined,
          "number": "+33  (8)2 06 92 96 57",
          "primary": true,
          "type": undefined,
        },
      ],
      "vendorId": "people/c2584501361615515329",
    },
  },
  Object {
    "doc": Object {
      "address": Array [
        Object {
          "city": "Greenville",
          "country": "Ã‰tats-Unis",
          "label": undefined,
          "postcode": "29601",
          "primary": true,
          "street": "46-198 E Elford St",
          "type": undefined,
        },
      ],
      "birthday": "1987-10-05",
      "company": undefined,
      "email": Array [
        Object {
          "address": "michelle.seabury@keysoft.me",
          "label": undefined,
          "primary": true,
          "type": undefined,
        },
        Object {
          "address": "mitch.neal@opticast.info",
          "label": undefined,
          "primary": false,
          "type": undefined,
        },
      ],
      "metadata": Object {
        "google": Object {
          "metadata": Object {
            "objectType": "PERSON",
            "sources": Array [
              Object {
                "etag": "#OzF7SvgKOWU=",
                "id": "33f916e00d9f5889",
                "type": "CONTACT",
                "updateTime": "2018-03-20T11:47:42.473001Z",
              },
            ],
          },
        },
        "version": 1,
      },
      "name": Object {
        "familyName": "Seabury",
        "givenName": "Michelle",
      },
      "note": undefined,
      "phone": Array [
        Object {
          "label": undefined,
          "number": "+33 (6)0 80 49 75 59",
          "primary": true,
          "type": undefined,
        },
        Object {
          "label": undefined,
          "number": "+33 (6)5 59 99 73 12",
          "primary": false,
          "type": undefined,
        },
      ],
      "vendorId": "people/c3745049716692310153",
    },
  },
  Object {
    "doc": Object {
      "address": Array [
        Object {
          "city": "Hampton",
          "country": "UK",
          "label": "Work",
          "postcode": undefined,
          "primary": true,
          "street": "78 High Street",
          "type": "work",
        },
      ],
      "birthday": "1912-06-23",
      "company": "National Physical Laboratory",
      "email": Array [
        Object {
          "address": "alan.turing@npl.com",
          "label": "Work",
          "primary": true,
          "type": "work",
        },
        Object {
          "address": "alan.turing@wanadoo.fr",
          "label": "Home",
          "primary": false,
          "type": "home",
        },
      ],
      "metadata": Object {
        "google": Object {
          "metadata": Object {
            "objectType": "PERSON",
            "sources": Array [
              Object {
                "etag": "#H6QA7yWPAHE=",
                "id": "7486d2d60ee594e7",
                "type": "CONTACT",
                "updateTime": "2018-04-03T15:16:02.276Z",
              },
            ],
          },
        },
        "version": 1,
      },
      "name": Object {
        "familyName": "Turing",
        "givenName": "Alan",
      },
      "note": undefined,
      "phone": Array [
        Object {
          "label": "Home",
          "number": "+44000555000",
          "primary": true,
          "type": "home",
        },
        Object {
          "label": "Work",
          "number": "+44 55 5000 555",
          "primary": false,
          "type": "work",
        },
      ],
      "vendorId": "people/c8396630372093629671",
    },
  },
  Object {
    "doc": Object {
      "address": Array [],
      "birthday": undefined,
      "company": "Winterfell",
      "email": Array [
        Object {
          "address": "eddard.stark@stark.com",
          "label": "Domicile",
          "primary": true,
          "type": "Domicile",
        },
      ],
      "metadata": Object {
        "google": Object {
          "metadata": Object {
            "objectType": "PERSON",
            "sources": Array [
              Object {
                "etag": "#jzax3Z1MggE=",
                "id": "257de03988942c82",
                "type": "CONTACT",
                "updateTime": "2018-04-03T15:22:30.530001Z",
              },
            ],
          },
        },
        "version": 1,
      },
      "name": Object {
        "familyName": "Stark",
        "givenName": "Eddard",
      },
      "note": undefined,
      "phone": Array [
        Object {
          "label": "Mobile",
          "number": "12 34 56 78 90",
          "primary": true,
          "type": "mobile",
        },
      ],
      "vendorId": "people/c2701561889201335426",
    },
  },
  Object {
    "doc": Object {
      "address": Array [
        Object {
          "city": undefined,
          "country": undefined,
          "label": undefined,
          "postcode": undefined,
          "primary": true,
          "street": "House Stark",
          "type": undefined,
        },
      ],
      "birthday": "0298-01-01",
      "company": "Night's Watch",
      "email": Array [
        Object {
          "address": "jon.snow@nightswatch.com",
          "label": "Work",
          "primary": true,
          "type": "work",
        },
      ],
      "metadata": Object {
        "google": Object {
          "metadata": Object {
            "objectType": "PERSON",
            "sources": Array [
              Object {
                "etag": "#jzax3Z1MggE=",
                "id": "59575f3f0a9cc5c8",
                "type": "CONTACT",
                "updateTime": "2018-04-03T15:22:30.530001Z",
              },
            ],
          },
        },
        "version": 1,
      },
      "name": Object {
        "familyName": "Snow",
        "givenName": "Jon",
      },
      "note": undefined,
      "phone": Array [],
      "vendorId": "people/c6437718916715038152",
    },
  },
  Object {
    "doc": Object {
      "address": Array [],
      "birthday": undefined,
      "company": "Faceless Men",
      "email": Array [
        Object {
          "address": "arya.stark@stark.com",
          "label": "Domicile",
          "primary": true,
          "type": "Domicile",
        },
        Object {
          "address": "agirl@noname.com",
          "label": "Work",
          "primary": false,
          "type": "work",
        },
      ],
      "metadata": Object {
        "google": Object {
          "metadata": Object {
            "objectType": "PERSON",
            "sources": Array [
              Object {
                "etag": "#jzax3Z1MggE=",
                "id": "2022cc1e8db59eb0",
                "type": "CONTACT",
                "updateTime": "2018-04-03T15:22:30.530001Z",
              },
            ],
          },
        },
        "version": 1,
      },
      "name": Object {
        "familyName": "Stark",
        "givenName": "Arya",
      },
      "note": "Beatae atque omnis nostrum eum. Quia sint ea a consequatur tempore et et animi. Ad quae perferendis sit qui aut fugiat deserunt.",
      "phone": Array [],
      "vendorId": "people/c2315637590020431536",
    },
  },
  Object {
    "doc": Object {
      "address": Array [],
      "birthday": undefined,
      "company": "Night's Watch",
      "email": Array [
        Object {
          "address": "samwel.tarly@nightswatch.com",
          "label": "Work",
          "primary": true,
          "type": "work",
        },
      ],
      "metadata": Object {
        "google": Object {
          "metadata": Object {
            "objectType": "PERSON",
            "sources": Array [
              Object {
                "etag": "#jzax3Z1MggE=",
                "id": "16c2056d0f8a078a",
                "type": "CONTACT",
                "updateTime": "2018-04-03T15:22:30.530001Z",
              },
            ],
          },
        },
        "version": 1,
      },
      "name": Object {
        "familyName": "Tarly",
        "givenName": "Samwel",
      },
      "note": undefined,
      "phone": Array [],
      "vendorId": "people/c1639879180286560138",
    },
  },
  Object {
    "doc": Object {
      "address": Array [],
      "birthday": undefined,
      "company": undefined,
      "email": Array [
        Object {
          "address": "noname@faceless.com",
          "label": "CoordonnÃ©es professionnelles",
          "primary": true,
          "type": "CoordonnÃ©es professionnelles",
        },
      ],
      "metadata": Object {
        "google": Object {
          "metadata": Object {
            "objectType": "PERSON",
            "sources": Array [
              Object {
                "etag": "#jzax3Z1MggE=",
                "id": "692105bd8e572536",
                "type": "CONTACT",
                "updateTime": "2018-04-03T15:22:30.530001Z",
              },
            ],
          },
        },
        "version": 1,
      },
      "name": Object {
        "familyName": undefined,
        "givenName": undefined,
      },
      "note": "Beatae atque omnis nostrum eum. Quia sint ea a consequatur tempore et et animi. Ad quae perferendis sit qui aut fugiat deserunt.",
      "phone": Array [
        Object {
          "label": undefined,
          "number": "00 0-555-000",
          "primary": true,
          "type": undefined,
        },
      ],
      "vendorId": "people/c7575342359908918582",
    },
  },
]
`;
